pkg/storage/ingest/reader.go

```diff
@@ -0,0 +1,331 @@
+	return r, nil
+}
+
+func (r *PartitionReader) start(ctx context.Context) error {
+	offset, err := r.fetchLastCommittedOffsetWithRetries(ctx)
+	if err != nil {
+		return err
+	}
+	level.Info(r.logger).Log("msg", "resuming consumption from offset", "offset", offset)
+
+	r.client, err = r.newKafkaReader(offset)
+	if err != nil {
+		return errors.Wrap(err, "creating kafka reader client")
+	}
+	r.admClient = kadm.NewClient(r.client)
+
+	return nil
+}
+
+func (r *PartitionReader) stop(error) error {
+	r.client.Close()
+	// r.admClient needs no closing since it's using r.client
+	return nil
+}
+
+func (r *PartitionReader) run(ctx context.Context) error {
+	defer r.commitLoopWg.Wait()
+
+	consumeCtx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	r.commitLoopWg.Add(1)
+	go r.commitLoop(consumeCtx)
+
+	for ctx.Err() == nil {
+		fetches := r.client.PollFetches(ctx)
+		if fetches.Err() != nil {
+			if errors.Is(fetches.Err(), context.Canceled) {
+				return nil
+			}
+			err := collectFetchErrs(fetches)
+			level.Error(r.logger).Log("msg", "encountered error while fetching", "err", err)
+			continue
+		}
+
+		r.recordFetchesMetrics(fetches)
+		r.consumeFetches(consumeCtx, fetches)
+		r.enqueueCommit(fetches)
+	}
+
+	return nil
+}
+
+func collectFetchErrs(fetches kgo.Fetches) (_ error) {
+	mErr := multierror.New()
+	fetches.EachError(func(s string, i int32, err error) {
+		// kgo advises to "restart" the kafka client if the returned error is a kerr.Error.
+		// Recreating the client would cause duplicate metrics registration, so we don't do it for now.
+		mErr.Add(err)
+	})
+	return mErr.Err()
+}
+
+func (r *PartitionReader) enqueueCommit(fetches kgo.Fetches) {
+	r.commitFetches <- fetches
+}
+
+func (r *PartitionReader) consumeFetches(ctx context.Context, fetches kgo.Fetches) {
+	records := make([]record, 0, len(fetches.Records()))
+
+	var minOffset, maxOffset int
+	fetches.EachRecord(func(record *kgo.Record) {
+		minOffset = min(minOffset, int(record.Offset))
+		maxOffset = max(maxOffset, int(record.Offset))
+		records = append(records, mapRecord(record))
+	})
+
+	err := r.consumer.consume(ctx, records)
+	if err != nil {
+		level.Error(r.logger).Log("msg", "encountered error processing records; skipping", "min_offset", minOffset, "max_offset", maxOffset, "err", err)
+		// TODO abort ingesting & back off if it's a server error, ignore error if it's a client error
+	}
+}
+
+func mapRecord(r *kgo.Record) record {
+	return record{
+		content:  r.Value,
+		tenantID: string(r.Key),
+	}
+}
+
+func (r *PartitionReader) recordFetchesMetrics(fetches kgo.Fetches) {
+	var (
+		now        = time.Now()
+		numRecords = 0
+	)
+
+	fetches.EachRecord(func(record *kgo.Record) {
+		numRecords++
+		r.metrics.receiveDelay.Observe(now.Sub(record.Timestamp).Seconds())
+	})
+
+	r.metrics.recordsPerFetch.Observe(float64(numRecords))
+}
+
+func (r *PartitionReader) newKafkaReader(at kgo.Offset) (*kgo.Client, error) {
```

Comment:
Not a blocker for this PR, but please add a TODO to your issue to review all consumer-related config options and check what should be fine-tuned. I got some good improvements fine-tuning the producer's ones in the Writer. Can be done in a follow up PR.